pkg/cmd/retlint/README.md

```diff
@@ -0,0 +1,86 @@
+# Return Linter
+
+> It would be nice if only `pgerror.Error` were actually returned from
+> an API, even though the functions are declared as returning `error`.
+
+```
+Output of retlint --help goes here.
+```
+
+## Motivation
+
+Due to the lack of return-type variance in Go's type system, it is
+often necessary to declare a return type that is "wider" than the
+actual types that will be returned.  A very common example, and the
+one in which the rest of the document is framed, is that we want to
+ensure that a concrete, user-facing error type, `pgerror.Error` is
+actually returned from functions that declare `error` as a return type.
+
+## Approach
+
+Thanks to the availability of the
+[`analysis`](https://godoc.org/golang.org/x/tools/go/analysis) and
+[`ssa`](https://godoc.org/golang.org/x/tools/go/ssa) packages, it is
+relatively straightforward to build a linter which can examine the call
+graph and type flow within a package or to perform a whole-world
+analysis on a program.
+
+The input to the linter is as follows:
+* A target interface, e.g. `error`.
+* A set of concrete types which implement the target interface,
+  e.g. `pgerror.Error`.
+* A filter which selects "seed" functions that must be provable to only
+  return the desired concrete types wherever the target interface is
+  declared to be returned.
+
+The linter will then perform an inductive, heuristic analysis, starting
+from the seed functions. The goal of this analysis is to classify each
+reachable function into those that always return one of the desired
+types in place of the target interface and those that do not.  In cases
+where a seed function cannot be proven to return only the desired
+concrete types, a lint error will be generated.  The lint error will
+include the seed function and at least one call-path chain to show the
+source of the undesirable type.
```

Comment:
you mention only two outcomes in this paragraph but the section below details 3 states. Was that intentional?